api/src/task/notifications/closing_date_notification.py

```diff
@@ -0,0 +1,193 @@
+import logging
+from datetime import timedelta
+from uuid import UUID
+
+from sqlalchemy import and_, exists, select, update
+from sqlalchemy.orm import joinedload
+
+from src.adapters import db
+from src.db.models.opportunity_models import Opportunity, OpportunitySummary
+from src.db.models.user_models import UserOpportunityNotificationLog, UserSavedOpportunity
+from src.task.notifications.base_notification import BaseNotification
+from src.task.notifications.constants import NotificationReason, UserEmailNotification
+from src.util import datetime_util
+
+logger = logging.getLogger(__name__)
+
+CONTACT_INFO = (
+    "mailto:support@grants.gov\n"
+    "1-800-518-4726\n"
+    "24 hours a day, 7 days a week\n"
+    "Closed on federal holidays"
+)
+
+
+class ClosingDateNotification(BaseNotification):
+
+    def __init__(
+        self,
+        db_session: db.Session,
+        frontend_base_url: str | None = None,
+    ):
+        super().__init__(db_session)
+
+        self.frontend_base_url = frontend_base_url
+
+    def collect_email_notifications(self) -> list[UserEmailNotification]:
+        """Collect notifications for opportunities closing in two weeks"""
+        two_weeks_from_now = datetime_util.utcnow() + timedelta(days=14)
+
+        # Find saved opportunities closing in two weeks that haven't been notified
+        stmt = (
+            select(UserSavedOpportunity)
+            .options(joinedload(UserSavedOpportunity.user))
```

Comment:
selectinload seems to work better - I've seen cases where joinedload just doesn't return for big enough queries. Might be safer to use that instead.